pandas/_libs/index.pyx

```diff
@@ -352,7 +352,7 @@ cdef class IndexEngine:
 
 cdef Py_ssize_t _bin_search(ndarray values, object val) except -1:
     cdef:
-        Py_ssize_t mid, lo = 0, hi = len(values) - 1
+        Py_ssize_t mid = 0, lo = 0, hi = len(values) - 1
```

Comment:
This is a real bug I think. mid is referenced in the return value, if len(values)=1 so the main loop immediately falls through.

Reply:
So this would fail if `len(values) == 0` right? Do you see a code sample to actually trigger that?

Reply:
quite possibly this is meant to be "caller checks". But as an independent module which that's out of the compiler's view.

Reply:
Yea sure. Similar comment though - good to suppress compiler warning but if there's an actual error here (which I think would still happen without the warning) would be good to address

Reply:
if values is empty, the loop falls through, `util.get_value_at(values, mid)` gets called when values is empty, which calls validate_indexer which raises. I think is ok.

Reply:
> which raises

cython code can have weird corner cases where exceptions get ignored.  Adding a test for this case seems worthwhile.

Reply:
interesting. example?

Reply:
huh. I think you've nailed something here. So it's not exactly weird corner  cases, but cdef function which return exception as if they were python.
https://cython.readthedocs.io/en/latest/src/userguide/language_basics.html#error-return-values

`validate_indexer` and `get_value_at` are both cdef. validate_indexer has ``except -1`, so any exception translates into a -1 retval. But the caller `get_value_at` doesn't check that and uses -1 to reference into the the empty array.

OTOH, the index constructor refuses to create empty indexes, so an end-to-end test does not seem possible. and  `get_value_at`  is used all over the place.

Hmm. will have to think about this.

Reply:
```python
In [18]: %%cython
    ...: 
    ...: cdef int foo() except -1:
    ...:     1/0
    ...:     cdef:
    ...:         int mid = 42
    ...:     return mid
    ...:     
    ...: cdef int bar():
    ...:     res = foo()
    ...:     return res
    ...:     
    ...: def baz():
    ...:     return bar()

In [19]: baz()
---------------------------------------------------------------------------
ZeroDivisionError                         Traceback (most recent call last)
_cython_magic_a591cd36cd8e49116d6e370949ee24c2.pyx in _cython_magic_a591cd36cd8e49116d6e370949ee24c2.foo()

ZeroDivisionError: float division
Exception ignored in: '_cython_magic_a591cd36cd8e49116d6e370949ee24c2.bar'
Traceback (most recent call last):
  File "_cython_magic_a591cd36cd8e49116d6e370949ee24c2.pyx", line 3, in _cython_magic_a591cd36cd8e49116d6e370949ee24c2.foo
ZeroDivisionError: float division
Out[19]: 0
```

very interesting bug.